test/runtests.jl

```diff
@@ -12,28 +19,84 @@ struct Foo
 end
 
 # test basic match
-x = nothing
-@test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
-@test x == 1
+let x = nothing
+    @test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
+    @test x == 1
+end
 
 # variables not bound if match fails
-x = nothing
-@test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
-@test x == nothing
+let x = nothing
+    @test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
+    @test x == nothing
+end
 
 # doesn't overwrite variables in outer scope
-x = nothing
-@test (@match Foo(1,2) begin
-    Foo(x,2) => x
-end) == 1
-@test x == nothing
+let x = nothing
+    @test (@match Foo(1,2) begin
+        Foo(x,2) => x
+    end) == 1
+    @test x == nothing
+end
 
 # variables not bound if guard fails
-x = nothing
-@test_throws MatchFailure @match Foo(1,2) begin
-  Foo(x, 2) where x != 1 => :ok
+let x = nothing
+    @test_throws MatchFailure @match Foo(1,2) begin
+      Foo(x, 2) where x != 1 => :ok
+    end
+    @test x == nothing
+end
+
+# match one struct field by name
+let x1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(x=x1) => x1
+    end) == 1
+end
+
+# match struct with mix of by-value and by-field name
+let x1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(0,2) => nothing
+           Foo(x=x1) => x1
+    end) == 1
+end
+
+# match multiple struct fields by name
+let x1 = nothing, y1 = nothing
+    @test (@match Foo(1,2) begin
+           Foo(x=x1,y=y1) => (x1,y1)
+    end) == (1,2)
+end
+
+# match struct field by name redundantly
+let x1 = nothing, x2 = nothing
+    @test_throws assertion_error @eval (@match Foo(1,2) begin
+           Foo(x=x1,x=x2) => (x1,x2)
+    end)
+end
+
+let x1 = 0
+    @test x1 == 0
+    @test (@match Foo(1,2) begin
+           Foo(x=x1, y=2) => (x1, 2)
+           end) == (1,2)
+    @test x1 == 0
```

Comment:
Oh sorry, i misunderstood the variable overwriting behavior when I suggested the test for `x1`'s value.

You can probably delete those extra lines around this if you want, since i think the non-overwriting behavior is already tested above. But up to you.